### PR TITLE
Refactor 1 of 2: Feature parity

### PR DESCRIPTION
**Pull request by exogen**
This contains some boilerplate-sync, some optimizations, lots of cleanup, lint fixes (based on latest lint rules), and uses `VictoryLabel` instead of `text`. The main purpose is to separate all this cleanup from the actual major feature change, which is using `Tick`, `GridLine`, etc. child components, and checking `role` for whether or not to render the default ones. That PR will be coming shortly.

/cc @boygirl @kenwheeler

### CODE REVIEW

File: demo/app.jsx [hunk 1]
```diff
@@ -65,7 +68,8 @@ class App extends React.Component {
             label={"animation\nwow!"}
             tickValues={this.state.tickValues}
             tickFormat={["first", "second", "third", "fourth", "fifth"]}
-            animate={{velocity: 0.01}}/>
+            animate={{velocity: 0.01}}
+          />
```
Comment (exogen) [inline demo/app.jsx:67]: Just because of a new lint rule that wants the `/>` to be dedented.

File: src/components/victory-axis.jsx [hunk 2]
```diff
@@ -507,22 +474,25 @@ export default class VictoryAxis extends React.Component {
...
+        "labelPadding", "offsetX", "offsetY", "padding", "width", "height"
       ]);
       return (
         <VictoryAnimation {...this.props.animate} data={animateData}>
-          {props => <VictoryAxis {...this.props} {...props} animate={null}/>}
+          {(props) => <VictoryAxis {...this.props} {...props} animate={null}/>}
         </VictoryAnimation>
       );
+    } else {
+      this.getCalculatedValues(this.props);
```
Comment (exogen) [inline src/components/victory-axis.jsx:487]: I removed `componentWillMount` and `componentWillReceiveProps` and put `getCalculatedValues` here. This aligns with advice I've seen surrounding computed values in React: calculate them at the latest possible time. Out of these three lifecycle methods, React is potentially going to call `render` the most sparingly.

File: src/components/victory-axis.jsx [hunk 3]
```diff
@@ -507,22 +474,25 @@ export default class VictoryAxis extends React.Component {
...
-      <g style={this.style.parent} transform={this.transform}>
-        {this.getGridLines()}
-        {this.getAxisLine()}
-        {this.getTickLines()}
-        {this.getLabelElements()}
+      <g style={this.style.parent} transform={transform}>
+        {this.renderGridLines()}
+        {this.renderAxisLine()}
+        {this.renderTicks()}
+        {this.renderLabel()}
```
Comment (exogen) [inline src/components/victory-axis.jsx:495]: This is a convention I think we should adopt, and one I've seen elsewhere: If we break out rendering into smaller methods, those methods should be named like `renderFoo` – that way it's clear when they get called and what they return.

File: src/components/victory-axis.jsx [hunk 4]
```diff
@@ -507,22 +474,25 @@ export default class VictoryAxis extends React.Component {
       // Do less work by having `VictoryAnimation` tween only values that
       // make sense to tween. In the future, allow customization of animated
       // prop whitelist/blacklist?
-      const animateData = _.omit(this.props, [
-        "orientation", "scale", "tickFormat", "animate",
-        "crossAxis", "standalone"
+      const animateData = _.pick(this.props, [
+        "style", "domain", "range", "tickCount", "tickValues",
+        "labelPadding", "offsetX", "offsetY", "padding", "width", "height"
```
Comment (exogen) [inline src/components/victory-axis.jsx:479]: Moving from `omit` to `pick` (the inverse) is a good idea in all our components, for animation performance. `pick` the smallest number of props, only the ones that make sense to animate.

File: src/components/victory-axis.jsx [hunk 5]
```diff
@@ -354,147 +351,117 @@ export default class VictoryAxis extends React.Component {
...
+          <line x2={props.x2} y2={props.y2} style={style.line}/>
+          <VictoryLabel
+            x={props.x}
+            y={props.y}
+            style={style.label}
+            textAnchor={props.textAnchor}
+            verticalAnchor={props.verticalAnchor}
+          >
+            {tickFormat.call(this, tick, index)}
+          </VictoryLabel>
```
Comment (exogen) [inline src/components/victory-axis.jsx:416]: This will be moved into at `Tick` component in the next PR.

File: src/components/victory-axis.jsx [hunk 6]
```diff
@@ -354,147 +351,117 @@ export default class VictoryAxis extends React.Component {
...
+      y1: this.padding.top,
+      y2: this.props.height - this.padding.bottom
+    } : {
+      x1: this.padding.left,
+      x2: this.props.width - this.padding.right
     };
-    return this.isVertical ?
-      <line y1={_.min(extent.y)} y2={_.max(extent.y)} style={this.style.axis}/> :
-      <line x1={_.min(extent.x)} x2={_.max(extent.x)} style={this.style.axis}/>;
+    return <line {...props} style={this.style.line}/>;
```
Comment (exogen) [inline src/components/victory-axis.jsx:392]: Will be moved to an `AxisLine` component.

File: src/components/victory-axis.jsx [hunk 7]
```diff
@@ -291,22 +294,21 @@ export default class VictoryAxis extends React.Component {
...
-      } else {
-        t = ticks;
+        return _.includes(ticks, 0) ? _.without(ticks, 0) : ticks;
       }
-    } else {
-      t = this.scale.domain();
+      return ticks;
     }
-    return Array.isArray(t) ? t : [t];
+    return this.scale.domain();
```
Comment (exogen) [inline src/components/victory-axis.jsx:311]: This `isArray` didn't seem to be doing anything, as I think all the `if` branches above returned arrays anyway. @boygirl Correct me if that's wrong!

File: src/components/victory-axis.jsx [hunk 8]
```diff
@@ -18,29 +26,40 @@ const defaultStyles = {
...
+  right: 1,
+  bottom: 1
+};
+
+const orientationVerticality = {
+  top: false,
+  bottom: false,
+  left: true,
+  right: true
+};
```
Comment (exogen) [inline src/components/victory-axis.jsx:61]: These are reused in various computations.

**Comment by boygirl** [comment]
cc/ @exogen I tried this out on it's own, and linked it and tried from chart, and it works well (some expected style differences). I'm not crazy about the changes to the style namespaces, but I'll reserve judgment until I can see how it all goes together in your next pr. Approved

**Comment by exogen** [comment]
@boygirl We could almost surely keep the `style` namespace the same as it was – but there's reasoning behind it. If `Tick` is its own subcomponent, for example, then `VictoryAxis` can pass `style.tick` along to that wholesale, and `Tick` then applies the `line` and `label` objects to the components _it_ renders. It keeps the meaning of `style` consistent: every key in `style` is applied to an element in your _shallow_ render. If one of those elements is a component with multiple elements itself, then its `style` object follows the exact same rule.

**Comment by boygirl** [comment]
The reasoning behind it makes sense, I'm just thinking about how it will feel to use api-wise. I think the shallower styles with unique namespaces are easier to read at a glance

**Comment by boygirl** [comment]
@exogen this looks ready to merge to me

**Pull request merged**